open_spiel/algorithms/infostate_tree.cc

```diff
@@ -0,0 +1,814 @@
+        child->sequence_id_ = SequenceId(sequence_index++, this);
+      }
+      // We could use sequence_index to set start and end sequences for
+      // the decision infostate right away here, however we'd like to make
+      // sure to label correctly all nodes in the tree.
+    }
+  }
+  // Finally label the last sequence (an empty sequence) in the root node.
+  sequences_.push_back(mutable_root());
+  mutable_root()->sequence_id_ = SequenceId(sequence_index, this);
+
+  CollectStartEndSequenceIds(mutable_root(), mutable_root()->sequence_id());
+}
+
+// Make a recursive call to assign the parent's sequences appropriately.
+// Collect pairs of (start, end) sequence ids from children and propagate
+// them up the tree. In case that deep nodes (close to the leaves) do not
+// have any child decision nodes, set the (start, end) to the parent sequence.
+// In this way the range iterator will be empty (start==end) and well defined.
+std::pair<size_t, size_t> InfostateTree::CollectStartEndSequenceIds(
+    InfostateNode* node, const SequenceId parent_sequence) {
+  size_t min_index = kUndefinedNodeId; // This is a large number.
+  size_t max_index = 0;
+  const SequenceId propagate_sequence_id =
+      node->sequence_id_.is_undefined()
+      ? parent_sequence
+      : node->sequence_id();  // This becomes the parent for next nodes.
+
+  for (InfostateNode* child : node->child_iterator()) {
+    auto[min_child, max_child] = CollectStartEndSequenceIds(
+        child, propagate_sequence_id);
+    min_index = std::min(min_child, min_index);
+    max_index = std::max(max_child, max_index);
+  }
+
+  if (min_index != kUndefinedNodeId) {
+    SPIEL_CHECK_LE(min_index, max_index);
+    node->start_sequence_id_ = SequenceId(min_index, this);
+    node->end_sequence_id_ = SequenceId(max_index + 1, this);
+  } else {
+    node->start_sequence_id_ = propagate_sequence_id;
+    node->end_sequence_id_ = propagate_sequence_id;
+  }
+
+  if (node->sequence_id_.is_undefined()) {
+    // Propagate children limits.
+    node->sequence_id_ = parent_sequence;
+    return {min_index, max_index};
+  } else {
+    // We have hit a defined sequence id, propagate it up.
+    return {node->sequence_id_.id(), node->sequence_id_.id()};
+  }
+}
+
+std::pair<double, SfStrategy> InfostateTree::BestResponse(
+    TreeplexVector<double>&& gradient) const {
+  SPIEL_CHECK_EQ(this, gradient.tree());
+  SPIEL_CHECK_EQ(num_sequences(), gradient.size());
+  SfStrategy response(this);
+
+  // 1. Compute counterfactual best response
+  // (i.e. in all infostates, even unreachable ones)
+  SequenceId current(0, this);
+  const double init_value = -std::numeric_limits<double>::infinity();
+  while (current.id() <= empty_sequence().id()) {
+    double max_value = init_value;
+    SequenceId max_id = current;
+    const InfostateNode* node = observation_infostate(current);
+    current = node->start_sequence_id();
+    while (current != node->end_sequence_id()) {
```

Comment:
This might be more readable as a `for` cycle, randomly adding a `continue` into the cycle would break it.